[RFC] SIMD implementation of vdev_raidz generate and reconstruct routines

ironMann opened the pull request:
based on top of @zfsonlinux:master 

This is a new implementation of RAIDZ1/2/3 routines using x86_64 scalar, SSE, and AVX2 instruction sets. Included are 3 parity generation routines (P, PQ, and PQR) and 7 reconstruction routines, for all RAIDZ level. On module load, a quick benchmark of supported routines will select the fastest for each operation and they will be used in runtime. Original implementation is still present and can be selected via module parameter.

Patch contains:

- specialized gen/rec routines for all RAIDZ levels,
- new scalar raidz implementation (unrolled for speed),
- two x86_64 SIMD implementations (SSE and AVX2 instructions sets),
- fastest routines selected on module load (benchmark).
- cmd/raidz_test tool for math exactness verification, code coverage and benchmark

New zfs module parameter:
```
- zfs_vdev_raidz_impl (string, can be changed online): selects the implementation to use
    "fastest" - use the fastest math available
    "original" - use the original raidz code
    "cycle" - cycle through all available impl for each new vdev_raidz (default in userspace tests)
    "scalar" - new scalar impl
    "sse" - new SSE impl if available
    "avx2" - new AVX2 impl if available
```

behlendorf commented:
Nice work.  It would be great to see this kind of optimization added to ZFS.  In order to make that happen it would be great if you (@ironMann) and @rdolbeau's, who has an implementation in #3374, could review each others work.  Then come up with a proposed patch set you're both happy with to review and start rigorously validating for correctness and performance.

Part of what's prevented this functionality from being being added in the need for some kind of torture test to verify it.  That's functionality which should be added to ztest.  Maybe something like have it randomly change the raidz implementation every few seconds.  Or even a dedicated test which verifies that all implementations generate the same parity.

ironMann commented:
I was not aware of previous efforts. I took a look at @rdolbeau's work. Basically, two implementations differ only in minor details. He has more hand inlined assembly, whereas I utilize smaller asm macros to achieve practically the same (both unroll 4x). I opted for this because it lets me have generic algorithm functions for all methods ([vdev_raidz_math_impl.h](https://github.com/ironMann/zfs/blob/wip-simd/module/zfs/vdev_raidz_math_impl.h) and [vdev_raidz_math_x86simd.h](https://github.com/ironMann/zfs/blob/wip-simd/module/zfs/vdev_raidz_math_x86simd.h)) that are using asm macros for different instruction sets (SSE and AVX2). I chose not to do AVX128 version because it lacks shuffle instructions needed for reconstruction methods, and also because that even SSE variant approaches memory throughput of a single CPU core. Lastly, I used non-temporal prefetch and streaming store instructions so that CPU caches are not trashed by the parity operations.
These are the differences I noticed, but performance wise, I bet they are pretty close.

Regarding testing, I have an external test tool that verifies bit exactness for all methods I wrote. I'll try to port it into ztest. I also left the 'old' implementation in, so switching in runtime between them is not a problem.
Suggestions and comments are welcome.

rdolbeau commented:
The 'good' branch for my code is abd2, since at the time the abd changes were meant to be merged. I don't know what the status is now, I haven't had time to follow ZFS developement closely.

abd2 also has simple benchmarking, AVX-512 (deactivated, as I've yet to be able to test if on actual hardware) & ARMv8/NEON. However, I don't have reconstruction and I didn't add explicit prefetch since I wasn't sure whether there was any data reuse between the parity code and other parts of ZFS.

I agree that AVX-128 is a bit redundant with SSE, since the kernel is probably no affected by the SSE/AVX transition (all SSE/AVX usage has to be explicit).

sempervictus commented:
I think the ABD changes are still meant to be merged, @tuxoko keeps improving the stack anyway, but it is kind of a problem for those of us building test systems with physical hardware for things like this (lots of rebuilds) that its still out of tree.

We still have a production host with SSE running the abd2 branch derived work, though i believe its from @tomgarcia's fork of @rdolbeau's work.

If there are advantages to this approach, whether in maintainability or performance, it would be nice to see work along this vein merged in sooner rather than later, as it's esoteric enough to become difficult to merge after any significant divergence upstream.

Thank you @ironMann and @rdolbeau for building out these low level solutions.

ironMann commented:
@behlendorf: I've added some code to use `ztest` for testing. Default behaviour for userspace now is to cycle through all supported implementations on each new zio. This should help with math stress testing. I'm running `ztest -r 11 -R 3 -a 0 -T 3600`. Maybe ztest should corrupt data more frequently? 
@rdolbeau: wow, AVX-512, that was preemptive :+1: I also agree with you. In any case, each inst set will be in a separate compilation unit, so tweaking it later for maximal performance should not be an issue.
@sempervictus: I'll bootstrap a couple of VM's to lighten the load on test servers. It's surprisingly not easy to keep up with all different platforms, distros, kernels, and compilers. As for performance, here is a kernel-mode benchmark (including fpu costs) from our test server: [results](https://gist.github.com/ironMann/9bd253a8024666a04c6a)

rdolbeau commented:
2016-02-18 17:31 GMT+01:00 Gvozden Neskovic <notifications@github.com>:
> @rdolbeau: wow, AVX-512, that was preemptive

With the masking and scatter/gather and CDI, AVX-512 seems to be "SIMD done
right at last" :-) There's two versions in the code, without (KNL) or with
(SKX) AVX512BW.

But they apparently forgot some instructions - there's not vpxorb in
AVX512BW. Someone must have thought that since there was just the one
variant in SSE and AVX, they didn't need more in AVX-512 since it's
bit-wise... completely forgetting about the masking :-( So we can't easily
do XOR-with-masking from a mask register. Too bad.

Cordially,

--
Romain Dolbeau

behlendorf commented:
@ironMann I know this is still a WIP but I made a few comments inline.  It looks like this is moving in a good direction, thanks for working on it.  I'm happy to make another pass perhaps once it's passing the buildbot.

> Maybe ztest should corrupt data more frequently? 

For testing purposes you could definitely increase the frequency of `ztest_fault_inject()` in `ztest.c`.

behlendorf commented:
To aid in the automated testing of these optimizations a [Intel Xeon E5-2676v3 (Haswell)](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20SIMD%20%28TEST%29) has been added to the TEST builders.  It supports the following cpu flags.

```
model name	: Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36
              clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl
              xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic
              movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor
              lahf_lm abm fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
```

ironMann commented:
@behlendorf @sempervictus I adapted the code to ABD patch (in #4439), but in the process, I had to reimplement all of the vector code to be able to consume HighMem pages efficiently. From this point on, I would prefer to concentrate on a single branch for further improvements. Basically the vector code is more or less stable on both branches, and now I want to work more on testing, verification and glue code for all implementation.
Any preference on which branch to proceed?

@behlendorf I seem that 2 zfstests on SIMD machine are consistently failing:
```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_scrub/zpool_scrub_002_pos (run as root) [00:01] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_scrub/zpool_scrub_003_pos (run as root) [00:02] [FAIL]
```

behlendorf commented:
@ironMann I'd recommend working on top of @tuxoko's work.  There's definitely ongoing work to get those changes in to state which can be maintained long term, but the intention is to move in that direction.

I suspect this is caused by a racy test case.  It looks like the scrub is either already finished by the time the test tries to stop it, or maybe it hasn't started yet.  Either way, I've seen this occasionally on other test instances and it's unrelated to these changes.  I'll open a new issue so we can track and fix the tests.

From the log:

```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_scrub/zpool_scrub_002_pos (run as root) [00:01] [FAIL]
18:37:37.51 ASSERTION: Verify scrub -s works correctly.
18:37:38.55 SUCCESS: /sbin/zpool scrub testpool.1405
18:37:38.55 cannot cancel scrubbing testpool.1405: there is no active scrub
18:37:38.55 ERROR: /sbin/zpool scrub -s testpool.1405 exited 1
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_scrub/zpool_scrub_003_pos (run as root) [00:02] [FAIL]
18:37:38.56 ASSERTION: scrub command terminates the existing scrub process and starts a new scrub.
18:37:39.81 SUCCESS: /sbin/zpool scrub testpool.1405
18:37:40.96 SUCCESS: /sbin/zpool scrub testpool.1405
18:37:40.96 zpool scrub don't stop existing scrubbing process.
```

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+				continue;
+			}
+
+			kpreempt_disable();
```
It's unsafe to disable preemption for these benchmarks because internally these tests cab perform sleeping allocations.  The kernel considers this to be an atomic context which triggers the usual BUG.  Since there isn't an easy way to make these allocations atomic I suggest dropping the `kpreempt_disable/enable()` and we'll have to live with the potentially precise benchmark results.

```
BUG: sleeping function called from invalid context at mm/slub.c:941
in_atomic(): 1, irqs_disabled(): 0, pid: 564, name: systemd-modules
```

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+
+			if (!curr_impl->is_supported()) {
+				continue;
+			}
```
[style] No braces around single line conditionals.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+			raidz_math_kstats[impl].gen_kstat[fn].value.ui64 =
+				speed;
+			bench_print(BENCH_RES_FMT, raidz_gen_name[fn],
+				curr_impl->name, speed/1000, speed%1000);
```
There a lot of redundant code here it would be good if you could refactor lines 393-414 and 440-458 somehow in to a generic `gen` and `rec` benchmark function.  If possible I'd also recommend adding the benchmark for the original  implementation after the `for (impl = 0; impl < impl_num; impl++) { }` loop.  That way you don't need to second pair of RAIDZ_GEN_NUM/RAIDZ_REC_NUM loops.  I think that would make this code much more readable.

ironMann replied:
I've refactored the whole benchmarking monstrosity, there's a lot less code duplication now.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+	/* Fake an zio and run the benchmark on it */
+	bench_zio = kmem_alloc(sizeof (zio_t), KM_SLEEP);
+	VERIFY(bench_zio);
+	bzero(bench_zio, sizeof (zio_t));
```
You can use `zio_root(NULL, NULL, NULL, 0)` to allocate this fake zio and then `zio_destroy()` to free it.

ironMann replied:
would need to make zio_destroy() non-static ?

behlendorf replied:
Ick, I hadn't noticed that.  Well then let's just tweak it to use `kmem_zalloc()` which will let you remove the VERIFY since the KM_SLEEP allocation can't fail and the `bzero()` since `kmem_zalloc()` will zero-fill.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+#define	BENCH_COLS	(3 + BENCH_D_COLS)
+#define	BENCH_COL_SIZE	(512)
+#define	BENCH_ZIO_SIZE	(BENCH_COL_SIZE * BENCH_D_COLS)
+#define	BENCH_NS	(1ULL*1000ULL*1000ULL) /* 50ms */
```
This is 1ms not 50ms as described in the comment.  You can use the `MSEC2NSEC()` macro for this conversion.

ironMann replied:
I'm not sure what is appropriate here. 
With 50ms, an AVX2 enabled server will spend 2s in benchmark ( 50ms * 10 routines * 4impl [orig, scalar, sse, avx2] ) 
blocking module load.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+
+			speed = run_count * (bench_zio->io_size >> 10);
+			speed /= (BENCH_D_COLS+bench_parity);
+			speed *= 1000000000ULL;
```
Use the `NANOSEC` macro here instead of `1000000000ULL`.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+			} while (gethrtime() < (t_start + BENCH_NS));
+
+			speed = run_count * (bench_zio->io_size >> 10);
+			speed /= (BENCH_D_COLS+bench_parity);
```
[style] add a space on either side of the `+`.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,713 @@
+#define	bench_print(x...)	cmn_err(CE_CONT, x)
+#else
+#define	bench_print(x...)	do { } while (0)
+#endif
```
Personally I don't think here's much benefit in printing this to the console, having access to this through the kstat seems sufficient to me so I'd suggest dropping `bench_print()` entirely.

behlendorf commented on tests/zfs-tests/tests/functional/raidz/raidz_001_neg.ksh:
```diff
@@ -0,0 +1,42 @@
+
+#
+# Copyright (c) 2016 by Gvozden Neskovic. All rights reserved.
+#
```
For entirely new files like this the Sun Microsystems copyright need not be included.  Also when there are multiple copyrights you can just list them as follows.

```
#
# Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
# Copyright (c) 2016 by Gvozden Neskovic. All rights reserved.
# Use is subject to license terms.
#
```

behlendorf replied:
Normally, I'm a big fan of splitting things up but in this case I think it would make good sense to squash this patch with the vectorized RAID implementation.

behlendorf commented on cmd/raidz_test/raidz_test.h:
```diff
@@ -0,0 +1,88 @@
+	return (a > 1 ? 1 + ilog2(a >> 1) : 0);
+}
+
+#define	PRINT(a...) (void) fprintf(stderr, a)
```
This would be a better place to define D_INFO and D_DEBUG.  It would allow them to be used in the IPRINT and DPRINT macros.

I'd also suggest dropping the PRINT macro and just using `fprintf()` directly for readability.  These messages should either be going to stdout if they're informational or stderr for a failure message.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+}
+
+static inline void
+__run_gen_bench(const char *impl)
```
[style] the existing code uses a `_impl()` suffix rather than a `__` prefix to denote a private wrapped function.  I'd recommend renaming as follows `run_gen_bench_impl()` for consistency.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+}
+
+static void
+__run_rec_bench(const char *impl)
```
[style] `__*` -> `*_impl`.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+	for (impl_name = (char **)raidz_impl_names; *impl_name != NULL;
+	    impl_name++) {
+
+		if (0 != vdev_raidz_impl_set(*impl_name)) {
```
[style] writing the conditional with the constant first as you've done is definitely is a safer way to write this code, but it's inconsistent with the rest of the code base.  For consistency it would be better to invert this check (and the others like it which have been added) to maintain the existing style.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+			elapsed = get_time_diff(&start, &stop);
+			disksize = (1ULL << ds) / rto_opts.rto_dcols;
+			d_bw = (double)iter_cnt * (double)(disksize);
+			d_bw /= (1024. * 1024. * elapsed);
```
[style] `1024.` -> `1024.0` for clarity.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+static double
+get_time_diff(struct rusage *start, struct rusage *stop)
+{
+	return (((double)stop->ru_utime.tv_sec*1e6 +
```
You learn something new every day, it never occurred to me to use 1e6.  That said, it would be clearer to use MICROSEC here.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+		    ddi_get_lbolt() + hz)) {
+
+			/* check if should stop the test (timeout) */
+			time_diff = (gethrtime() - start_time) / 1000000000ULL;
```
`s/1000000000/NANOSEC`

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	raidz_test_opts_t *opts;
+	int a, d, o, s;
+
+	max_free_slots = free_slots = MAX(2, sysconf(_SC_NPROCESSORS_ONLN));
```
FYI `boot_ncpus` could be used for `sysconf(_SC_NPROCESSORS_ONLN)` but I'm OK with either since this code is userspace only.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	return (NULL);
+}
+
+#define	ARR_LEN(x)	(sizeof (x) / sizeof (0[x]))
```
ARRAY_SIZE is already defined and available in user and kernel space and should be used instead.

behlendorf commented on cmd/raidz_test/raidz_bench.c:
```diff
@@ -0,0 +1,248 @@
+			elapsed = get_time_diff(&start, &stop);
+			disksize = (1ULL << ds) / rto_opts.rto_dcols;
+			d_bw = (double)iter_cnt * (double)disksize;
+			d_bw /= (1024. * 1024. * elapsed);
```
[style] /s/1024./1024.0/

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	bcopy(&rto_opts_defaults, o, sizeof (*o));
+
+	while ((opt = getopt(argc, argv,
+	    "TDBSvha:o:d:s:t:")) != EOF) {
```
The `getopt` return code should be checked against -1 not EOF.   They're not always the same and the getopt man page explicitly says `-1` is the value on error.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+		case 't':
+
+			value = strtoull(optarg, &end, 0);
+		}
```
This is a little non-standard, why not just move this optarg handling in to each case below?

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+
+		switch (opt) {
+		case 'a':
+			o->rto_ashift = MIN(12, MAX(9, value));
```
The utilities limit the ashift to 13 not 12.  It could go slightly higher be we start running out of space in the uberblock ring.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+			o->rto_dcols = MIN(255, MAX(1, value));
+			break;
+		case 's':
+			o->rto_dsize = 1ULL <<  MIN(24, MAX(9, value));
```
/s/9/SPA_MINBLOCKSHIFT
/s/24/SPA_MAXBLOCKSHIFT

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	init_zio_data(opts->zio_golden);
+	init_zio_data(zio_test);
+
+	VERIFY0(vdev_raidz_impl_set("original"));
```
Just FYI since this is always built with -DDEBUG you could use ASSERT instead since they'll always be promoted to VERIFY.  No need to change anything but something I wanted to mention.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+
+	*zio = umem_alloc(sizeof (zio_t), UMEM_NOFAIL);
+
+	bzero(*zio, sizeof (zio_t));
```
Could be simplified by using `umem_zalloc()`

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	if (err) {
+		PRINT("ERROR!\n");
+		exit(-1);
+	}
```
How can this ever happen?  This conditional and `err` variable look like they should be removed entirely.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	return (err);
+}
+
+unsigned
```
`err` is defined as an `int` here so I'd suggest changing this return type to match.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	for (d = 0; d < ARR_LEN(dcols_v); d++) {
+
+		if ((size_v[s] < (1<<ashift_v[a]) * o) ||
+		    (size_v[s] < (1<<ashift_v[a]) * dcols_v[d])) {
```
[style] add whitespace around `<<` operator.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+
+		VERIFY3P(thread_create(NULL, 0,
+		    (thread_func_t) sweep_thread, (void *) opts,
+		    0, TS_RUN, 0, PTHREAD_CREATE_JOINABLE),
```
These thread arguments are in the wrong order and the priority needs to be specified.  Something like this.

```
VERIFY3P(thread_create(NULL, 0, sweep_thread, opts, 0, curproc, TS_RUN, defclsyspri))
```

You'll also want to update the `sweep_thread()` prototype to return `void` so it matches the `thread_func_t`.  There's no need to return NULL from the thread function.

behlendorf commented on cmd/raidz_test/raidz_test.c:
```diff
@@ -0,0 +1,800 @@
+	action.sa_flags = 0;
+
+	if (sigaction(SIGSEGV, &action, NULL) < 0) {
+		(void) fprintf(stderr, "ztest: cannot catch SIGSEGV: %s.\n",
```
s/ztest/raidz_test/g

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,626 @@
+typedef void (*benchmark_fn)(raidz_map_t *rm, const int fn);
+
+static void
+__benchmark_gen(raidz_map_t *rm, const int fn)
```
Same as above private `__*` functions should be renamed `*_impl` for consistency.

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,626 @@
+	}
+
+	/* Benchmark data reconstruction methods */
+	bench_rm = vdev_raidz_map_alloc(bench_zio, 9, BENCH_COLS, PARITY_PQR);
```
Thanks for cleaning this up!  The benchmarking code is much better after you refactored it!

behlendorf commented on module/zfs/vdev_raidz_math.c:
```diff
@@ -0,0 +1,626 @@
+#define	BENCH_COLS	(BENCH_D_COLS + PARITY_PQR)
+#define	BENCH_COL_SIZE	(1ULL << 12)
+#define	BENCH_ZIO_SIZE	(BENCH_COL_SIZE * BENCH_D_COLS)
+#define	BENCH_NS	MSEC2NSEC(25)			/* 25ms */
```
So at 25ms per micro-benchmark and 40 benchmarks we're adding 1s to the module load time.  It makes me wonder if  it's worth only running the benchmarks when `zfs_vdev_raidz_impl == IMPL_FASTEST` (the default).

ironMann replied:
Benchmark could be moved in a separate thread, so it doesn't block module loading.
IMPL_FASTEST is currently the default.

behlendorf replied:
That's a thought, but let's leave it as is for now.  If it ever becomes an issue we can always optimize it then.

behlendorf commented:
@ironMann despite all my review comments this looks great!  Largely just a little bit of cleanup here and there remaining.

We've installed this code on a small 32 node development cluster for testing.  Our plan is to put it under a real workload and case real failures for a week or two.  We'll definite stress all the vectorized implementations and verify the reconstruction works correctly.  But if you can suggest any additional torture tests you'd like us to run with real hardware let me know.  

Assuming everything holds together, we don't uncover anything surprising, the above review comments are resolved, and your happy with the final implementation I'll be comfortable merging this to master.

ironMann commented:
@behlendorf I'll find some time until next week to work on improving the code, making it more consistent,  add more comments, etc.

As for testing, I don't anticipate any major (unsolvable) problems. Code coverage with unit tests was perfect 100% for all new RAID-Z code.  Hopefully, there should not be any surprises with running on the real hardware.

ironMann commented on module/zfs/vdev_raidz.c:
```diff
@@ -1739,10 +1633,13 @@ raidz_parity_verify(zio_t *zio, raidz_map_t *rm)
 /*
  * Keep statistics on all the ways that we used parity to correct data.
  */
 static uint64_t raidz_corrected[1 << VDEV_RAIDZ_MAXPARITY];
```
@behlendorf I found this statistics that's otherwise unused. Not really sure how useful it is.
Any suggestion what to do with it?

behlendorf replied:
@ironMann I'd suggest removing it, to my knowledge it's never proven useful in the past.  At time point we might go through and add some useful kstats if we find we need that kind of visibility in to the raidz code.

ironMann commented:
@behlendorf I've addressed your remarks, and reworked a part of reconstruction code to properly work with up to 255 child devices in raidz2 and raidz3 (limitation of the math).

I've noticed an issues with builds on non-x86_64 platforms. It seems that the userspace libs are being built without any optimizations, while requesting stack-frame to be below 1024B. This check fails for reconstruction methods because the code rely on function inlining. Since gcc does not produce sane stack size values without at least -O1 these warning get promoted to errors in DEBUG builds. see [build log](http://build.zfsonlinux.org/builders/Ubuntu%2014.04%20i686%20%28BUILD%29/builds/5171/steps/shell_3/logs/make)
Stack frame of this method is 280B on x64 build with -O2 opt level. The error can be triggered when -O0 is passed to CFLAGS.

behlendorf commented:
@ironMann thanks for refreshing this and addressing those comments.  It also looks like you've also fixed the build issue on non-x86_64 platforms?

The patch has been holding up very well in our testing.  The only issue we've encountered, and I haven't looked in to yet, is for some reason after loading the modules **original** is selected even though **avx2** has by far the best micro-benchmark results.  This is  with a version of the patch which predates your benchmark cleanup so perhaps you've already fixed this?

ironMann commented:
@behlendorf I've made simpler scalar routines, so they're not triggering the stack-limit error any more.

The parameter should read [fastest]  on init (unless cycle, original or scalar provided as module option). After benchmark routine is finished, other supported impl should turn up in the parameter file
```
# cat /sys/module/zfs/parameters/zfs_vdev_raidz_impl
[fastest] original cycle scalar sse avx2
```

'fastest' is intended to cherry pick the best routine it finds in the benchmarks, so it can be 'created' from more than one implementation.

ofaaland commented:
@ironMann I see the manual page zfs-module-parameters.5 states that, avx2 is not an allowed value for module parameter zfs_vdev_raidz_impl at module load time. However, if the user selects "fastest", zfs may choose to use avx2 (and it does on my test system).

What is the reason for this restriction?

Thanks for the work!

ironMann commented:
@ofaaland Basically, there's no way of knowing if vectorized implementations are supported before module is loaded. Thus, only scalar, original, and 'fastest' are allowed at load time. 
It would be too inconsistent to default to something else if requested impl is actually not available...

"fastest" is the default, and it will usually pick the highest vector methods, so there's no need to specify that parameter manually to get best performance.

ofaaland commented:
@ironMann thanks for the explanation.  Do you allow those options for testing purposes?  Or are they there so the user can choose the implementation, for cases when vdev_raidz_math_init() chooses an implementation that is problematic?

ironMann commented:
@ofaaland Well, initially it was for testing, but then it morphed to 'freedom of choice' kind of thing. I'm still not sure if 'cycle' option should be given to users (it is only useful for unit and coverage testing). Maybe it should be removed in non-debug builds.

In theory, `vdev_raidz_math_init()` could choose slower method due to bunch of reasons, but that is highly unlikely. But not, problematic implementation, in sense that it would not work correctly.

behlendorf commented:
@ironMann having 'cycle' available for testing has been helpful to me to build confidence in the patch.  But now that most of the testing done I wouldn't object to removing it as a option from the kernel module.  We'll of course want to leave it in user space for ztest though.

Thus far I haven't observed any problems with the patch.  I'll leave my testing running over the weekend and if I don't encounter any problems I'll sign off on this patch so it can be merged.  That just leaves the very minor issue of removing 'cycle' as mentioned above.

ofaaland commented:
@ironMann, feel free to reject it, but how about this text for the manpage?
`
Parameter for selecting raidz implementation to use.

Options marked (always) below may be selected on module load as they are
supported on all systems.

The remaining options may only be set after the module is loaded, as they
are available only if the implementations are compiled in and supported
on the running system.

Once the module is loaded, the content of
/sys/module/zfs/parameters/zfs_vdev_raidz_impl will show available options
with the currently selected one enclosed in [].

Possible options are:
  fastest  - (always) implementation selected using built-in benchmark
  original - (always) original raidz implementation
  scalar   - (always) scalar raidz implementation
  sse      - implementation using SSE instruction set (64bit x86 only)
  avx2     - implementation using AVX2 instruction set (64bit x86 only)
.sp
`

ironMann commented:
@ofaaland thanks, it's more clear.

@behlendorf In addition, I would also skip benchmark in userspace, and use the highest supported impl as the 'fastest'. This avoids startup delay of tools that call `kernel_init()`?

behlendorf commented:
@ironMann the `kernel_init()` optimization is a nice idea even if it should really only impact ztest.  It also looks like `vdev_zaps_005_pos` is failing on all the testers so that's going to need to be explained.

rdolbeau commented:
2016-06-18 12:53 GMT+02:00 Gvozden Neskovic <notifications@github.com>:
> @behlendorf In addition, I would also skip benchmark in userspace, and
use the highest supported impl as the 'fastest'. This avoids startup delay
of tools that call kernel_init()?

Which is the fastest might be obvious on X86-64 (it's likely
AVX512>AVX2>SSE>scalar), but for other architectures it's not necessarily
the case. There's a lot of e.g. Aarch64 cores out there, and it's possible
a wide-issue scalar core would outperform a 64-bits wide NEON pipe in some
cases. And the NEON itself might need to be tuned differently for different
cores.

Cordially,

--
Romain Dolbeau

ironMann commented:
@rdolbeau You are right, but there's currently no real use-case for having 'real fastest' impl in userspace tools. These apps cycle through all supported impl. for better test coverage. Real benchmark is still performed on 'zfs' module load, but removing it in userspace saves a lot of time on startup of `ztest` and `zdb`.

@behlendorf `vdev_zaps_*` seem to lack proper setup. If I run them standalone on my dev VM `005` fails, but they all pass in the whole suite. I've messed with the balance by adding `default_cleanup` in `/raidz/raidz_*`. I've tried adding `default_setup_noexit ${DISKS%% *}` to `zaps/setup.ksh` but that caused `zaps_001` to fail, and all others to pass... Sorry, I don't have any more time now to debug this further.

behlendorf commented:
@ironMann thanks for looking at the `vdev_zaps_*` failures.  It's clear these failures are unrelated to this PR and the test cases themselves need to be improved.  Given that they're transient and seem to be infrequent I'm OK with tackling them as their own issue.  So they don't need to hold this up.

The testing I've done over the last week on real hardware has been as abusive as I could make it, and it was designed to cover as much of the testing space as possible: raidz[123] geometries, 4k-16M block sizes,  1-20 device-per-raidz groups, full device rebuilds, all new algorithms checked.  I wasn't able to uncover any problems with the patch.

This patch LGTM.  Nice job!  Let me know if your happy with this as a final version so it can be merged.  We can always tweak the administrative aspects of it a little if needed after it's merged.  But the core of it looks completely solid.

ironMann commented:
@behlendorf Thanks, I think the patch is in a ok shape, too. The core raidz framework should be generic enough to support addition of new instruction set implementation with minimal effort (also 32bit SSE/AVX2 variants if relevant).

behlendorf commented:
Merged to master as:

ab9f4b0 SIMD implementation of vdev_raidz generate and reconstruct routines

thegreatgazoo commented:
I'm testing master branch today. A few comments/questions on this patch:
```
# cat /sys/module/zfs/parameters/zfs_vdev_raidz_impl
[fastest] original scalar sse avx2
```
So I'm using the fastest option, but which one is that? As an end-user, I'd want to know which implementation is actually in use, but "fastest" doesn't give me that information.


```
# ls -l /sys/module/zfs/parameters/zfs_vdev_raidz_impl
-rw-r--r-- 1 root root 4096 Jul 12 17:15 /sys/module/zfs/parameters/zfs_vdev_raidz_impl
# echo original > /sys/module/zfs/parameters/zfs_vdev_raidz_impl
-bash: echo: write error: Invalid argument
# echo "original" > /sys/module/zfs/parameters/zfs_vdev_raidz_impl
-bash: echo: write error: Invalid argument
```
The file is writable so I guess I'd be able to change it at run-time, but it ended up with write errors. Did I do something wrong here?

thegreatgazoo commented on man/man5/zfs-module-parameters.5:
```diff
@@ -1677,6 +1677,33 @@ Default value: \fB4,096\fR.
+\fBzfs_vdev_raidz_impl\fR (string)
+.ad
+.RS 12n
+Parameter for selecting raidz implementation to use.
```
I'd rather call it **raidz parity** implementation, just to be clear.

thegreatgazoo commented on module/zfs/vdev_raidz.c:
```diff
@@ -579,6 +457,10 @@ vdev_raidz_map_alloc(zio_t *zio, uint64_t unit_shift, uint64_t dcols,
 	zio->io_vsd_ops = &vdev_raidz_vsd_ops;
+
+	/* RAIDZ ops init */
+	vdev_raidz_math_get_ops(rm);
```
I think it'd be clearer:
```C
rm->rm_ops = vdev_raidz_math_get_ops();
```

ironMann replied:
agree

thegreatgazoo commented on module/zfs/vdev_raidz.c:
```diff
@@ -1436,33 +1323,40 @@ vdev_raidz_reconstruct(raidz_map_t *rm, int *t, int nt)
+	/*
 	 * See if we can use any of our optimized reconstruction routines.
 	 */
-	if (!vdev_raidz_default_to_general) {
```
Just curious, why was this conditional removed?

ironMann replied:
I wasn't sure why it was there, since it's not wired to any parameter, and it's technically unitialized (not static).

thegreatgazoo replied:
It was a file-scope global, so compiler would init it to 0. It doesn't have to have that _static_.

On systems other than Linux, I believe there's mechanisms to change such variables at run-time without explicitly declaring them as parameters.

I'm not arguing whether to keep this change. I just wanted to understand why it was made, since the description of this patch didn't mention anything about it.

ironMann commented:
@thegreatgazoo The fastest impl. is envisioned to be found at runtime. So there's 10 raidz parity routines in total, and in theory the fastest can pick up function pointers from all supported implementations, if that is preferable. But in practice, it's probably going to use the widest SIMD implementation your CPU supports.

> The file is writable so I guess I'd be able to change it at run-time, but it ended up with write errors. Did I do something wrong here?

That's news. Can you try with `printf "original" > /sys/module/zfs/parameters/zfs_vdev_raidz_impl`
It might be that `echo` adds a newline.

behlendorf commented:
> It might be that echo adds a newline.

This is exactly what's happening, we'll want to trim the trailing white space.  I definitely tested this so I'm not sure how I missed it.

ironMann commented:
@behlendorf It seems that `echo` is inconsistent with that. I'll fold these fixes into #4815

rdolbeau commented:
2016-07-12 20:05 GMT+02:00 Gvozden Neskovic <notifications@github.com>:
> But in practice, it's probably going to use the widest SIMD
implementation your CPU supports.

... on x86-64. On Aarch64, it depends on the core. Some have fast, wide
NEON support... some don't :-(

Cordially,

--
Romain Dolbeau

thegreatgazoo commented:
@ironMann Yes newline was the problem, _echo -n_ worked. Since mere _echo_ works with setting other parameters under _/sys/module/zfs/parameters/_, I'd think it should be fixed. BTW, I saw in the code:
```
module_param_call(zfs_vdev_raidz_impl, zfs_vdev_raidz_impl_set,
        zfs_vdev_raidz_impl_get, NULL, 0644);
```

The zfs_vdev_raidz_impl_get() grabs _vdev_raidz_impl_lock_, but zfs_vdev_raidz_impl_set() does not - that doesn't seem right to me.

thegreatgazoo commented:
@ironMann Thanks the explanation on **[fastest]**. My real question is, as the admin, I want to know **exactly** what that **[fastest]** is, but /sys/module/zfs/parameters/zfs_vdev_raidz_impl doesn't tell me that. Another scenario would be bug reporting, if a user suspects/hits a bug in the parity routine, we'd at least need to know what that **[fastest]** points to, preferrably without patching and rebooting since technically **[fastest]** is dynamic.

ironMann commented:
@thegreatgazoo That locking is indeed a bug. Thing is, that method can be called before the lock is even initialized (when module parameter is specified), and also through api from userspace. will fix.

That's a legitimate point about [fastest] option. Currently there's a kstat for measured throughput of all methods, and contents of [fastest] can be deduced from that, but there should exist a better way. I can extend kstat data to explicitly show what instruction set is used.
